extensions/azure/assetindex-cosmos/src/main/java/org/eclipse/dataspaceconnector/assetindex/azure/CosmosAssetIndexExtension.java

```diff
@@ -0,0 +1,114 @@
+/*
+ *  Copyright (c) 2020, 2021 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.assetindex.azure;
+
+import com.azure.cosmos.ConsistencyLevel;
+import com.azure.cosmos.CosmosClient;
+import com.azure.cosmos.CosmosClientBuilder;
+import com.azure.cosmos.CosmosContainer;
+import com.azure.cosmos.CosmosDatabase;
+import com.azure.cosmos.models.CosmosDatabaseResponse;
+import net.jodah.failsafe.RetryPolicy;
+import org.eclipse.dataspaceconnector.assetindex.azure.model.AssetDocument;
+import org.eclipse.dataspaceconnector.common.string.StringUtils;
+import org.eclipse.dataspaceconnector.cosmos.azure.AbstractCosmosConfig;
+import org.eclipse.dataspaceconnector.spi.EdcException;
+import org.eclipse.dataspaceconnector.spi.asset.AssetIndex;
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.eclipse.dataspaceconnector.spi.security.Vault;
+import org.eclipse.dataspaceconnector.spi.system.ServiceExtension;
+import org.eclipse.dataspaceconnector.spi.system.ServiceExtensionContext;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Provides a persistent implementation of the {@link org.eclipse.dataspaceconnector.spi.asset.AssetIndex} using CosmosDB.
+ */
+public class CosmosAssetIndexExtension implements ServiceExtension {
+
+    private static final String NAME = "CosmosDB Asset Index";
+
+    private Monitor monitor;
+
+    @Override
+    public Set<String> provides() {
+        return Set.of(AssetIndex.FEATURE);
+    }
+
+    @Override
+    public void initialize(ServiceExtensionContext context) {
+        monitor = context.getMonitor();
+        monitor.info(String.format("Initializing %s extension...", NAME));
+
+        var configuration = new AssetIndexCosmosConfig(context);
+        Vault service = context.getService(Vault.class);
+
+        // create client
+        var cosmosClient = createClient(service, configuration.getAccountName(), Collections.singletonList(configuration.getPreferredRegion()));
+
+        // get database, throw exception if not exists
+        var database = createDatabase(configuration, cosmosClient);
+        String containerName = configuration.getContainerName();
+        if (database.readAllContainers().stream().noneMatch(sp -> sp.getId().equals(containerName))) {
+            //todo: maybe create database if not exist?
```

Comment:
I assume in the most cases the EDC connects to an existing CosmosDB instance. Therefore throwing an exception is perfectly fine here. In case the database must be created together with the EDC application, using something like Terraform is probably better suited anyway.
Therefore removing this comment should be fine.

Reply:
Hi Dominik, 
agreed, I removed this comment.